upgradability/forward-call/address_book/lib.rs

```diff
@@ -0,0 +1,71 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use ink_lang as ink;
+
+#[ink::contract]
+mod address_book {
+    use ink_storage::{traits::SpreadAllocate, Mapping};
+    use ink_prelude::string::String;
+
+    const MAX_INFO_LEN: usize = 20;
+
+    #[ink(storage)]
+    #[derive(SpreadAllocate)]
+    pub struct AddressBook {
+        contact_info: Mapping<AccountId, String>,
+    }
+
+    impl AddressBook {
+        /// Initializes an empty contact info DB.
+        #[ink(constructor)]
+        pub fn new() -> Self {
+            ink_lang::utils::initialize_contract(|_| {})
+        }
+
+        /// Sets contact info of the caller.
+        /// Returns true if this operation was successful.
+        #[ink(message)]
+        pub fn set_info(&mut self, info: String) -> bool {
+            if info.len() > MAX_INFO_LEN || info.capacity() > MAX_INFO_LEN {
```

Comment:
Is checking`String::capacity` required? It is possible to have equal strings but[ with different capacities,](https://doc.rust-lang.org/std/string/struct.String.html#method.capacity), depending on how `String` is created.

Reply:
If we just check size, then we might get a String with large capacity (eg. with the use of `reserve`), and we probably do not want to keep that in our storage.

Reply:
But checking `len` is redundant here, I think.

Reply:
OK, good point